firedrake/checkpointing.py

```diff
@@ -815,14 +899,36 @@ def save_function(self, f, idx=None, name=None):
                 _name = "_".join([PREFIX_EMBEDDED, f.name()])
                 _f = Function(_V, name=_name)
                 self._project_function_for_checkpointing(_f, f, method)
-                self.save_function(_f, idx=idx)
+                self.save_function(_f, idx=idx, timestepping_info=timestepping_info)
                 self.set_attr(path, PREFIX_EMBEDDED + "_function", _name)
             else:
                 # -- Save function topology --
                 path = self._path_to_function(tmesh.name, mesh.name, V_name, f.name())
                 self.require_group(path)
                 self.set_attr(path, PREFIX + "_vec", tf.name())
                 self._save_function_topology(tf, idx=idx)
+                # store timstepping_info only if in timestepping mode
+                if idx is not None:
+                    path = self._path_to_function_timestepping(tmesh.name, mesh.name, V_name, f.name())
+                    new = path not in self.h5pyfile
+                    self.require_group(path)
+                    # We make sure the provided timestepping_info is consistent all along timestepping.
+                    if not new:
+                        existing_keys = {key.replace(PREFIX_TIMESTEPPING_HISTORY + "_", "", 1)
+                                         for key in self.h5pyfile[path].attrs.keys()
+                                         if key.startswith(PREFIX_TIMESTEPPING_HISTORY)}
+                        if timestepping_info.keys() != existing_keys:
+                            raise RuntimeError(
+                                r"Provided keys in timestepping_info must remain consistent")
+                    # store items in timestepping_info accordingly
+                    for ts_info_key, ts_info_value in timestepping_info.items():
+                        try:
+                            ts_info_value = float(ts_info_value)
+                        except ValueError:
+                            raise ValueError(f"{ts_info_key} should be convertible to float.")
```

Comment:
```suggestion
                        if not isinstance(ts_info_value, float):
                            raise NotImplementedError(f"timestepping_info must have float values: got {type(ts_info_value)} )
```
Let us restore what you originally did so that `int` will not be converted to `float` silently for now.